lib/SILGen/SILGen.cpp

```diff
@@ -823,7 +823,8 @@ void SILGenModule::emitDifferentiabilityWitness(
   auto *diffWitness = SILDifferentiabilityWitness::createDefinition(
       M, originalFunction->getLinkage(), originalFunction, loweredParamIndices,
       config.resultIndices, config.derivativeGenericSignature,
-      /*jvp*/ nullptr, /*vjp*/ nullptr, originalFunction->isSerialized(),
+      /*jvp*/ nullptr, /*vjp*/ nullptr,
+      /*isSerialized*/ hasPublicVisibility(originalFunction->getLinkage()),
```

Comment:
We probably shouldn't always serialize witnesses for any public declarations, because things we emit should be opaque by default.  Maybe you should add logic to the cross module optimizer that was recently added (#28407) to make differentiability witnesses serialized.

Reply:
I'll do this later because this PR is now very tested and almost ready to go. https://bugs.swift.org/browse/TF-1035

Reply:
Sounds good.  Thanks.